simphony/cuds/model.py

```diff
@@ -39,119 +55,179 @@ def __init__(self):
         # that will return the corresponding value.
         self._map = {}
 
-    @staticmethod
-    def _is_cuds_component(obj):
-        """Check whether the object is a CUDS component."""
-        # When I see a bird that walks like a duck and swims like a duck
-        # and quacks like a duck, I call that bird a duck
-        # TODO: replace with metadata type checks
-        return all([hasattr(obj, 'name'),
-                    hasattr(obj, 'data')])
+        # Another map to keep a mapping between names and ids
+        self._name_to_id_map = {}
+
+    @property
+    def uid(self):
+        return self._uid
+
+    @uid.setter
+    def uid(self, value):
+        raise TypeError("uid is readonly.")
+
+    @uid.deleter
+    def uid(self):
+        raise TypeError("Can't delete uid.")
 
     @staticmethod
     def _is_dataset(obj):
         """Check if the object is a dataset."""
-        return isinstance(obj, (ABCParticles, ABCLattice, ABCMesh))
+        return isinstance(obj, (ABCParticles,
+                                ABCLattice,
+                                ABCMesh,
+                                api.DataSet))
 
     @property
     def data(self):
         """Data container for CUDS items."""
         return DataContainer(self._data)
 
     @data.setter
-    def data_setter(self, value):
+    def data(self, value):
         self._data = DataContainer(value)
 
+    @data.deleter
+    def data(self):
+        raise TypeError("Can't delete data.")
+
     def add(self, component):
         """Add a component to the CUDS computational model.
 
-        If the component has `uid` attribute but that value is
-        None, a new uuid will be created and assigned to the
-        component after the component is successfully added
-        to the CUDS computational model.
+        This will replace any existing value with the same `uid`
```

Comment:
Should we add an `update` method so we keep consistent with the rest of simphony and we keep functionality separated?